### PR TITLE
improve default `AUTOMATIC_REPLACEMENTS`

### PR DESCRIPTION
**Pull request by ericphanson**
Now we normalize out multiple spaces to be a single space, and make the regex more robust to multiple replacements, and add some much-needed tests.

Thanks to @ararslan for pointing out issues with this.

### CODE REVIEW

File: Project.toml [hunk 1]
```diff
@@ -1,7 +1,7 @@
 name = "KeywordSearch"
 uuid = "f977ba8c-7144-47e8-b06b-e3f658952959"
 authors = ["Beacon Biosignals, Inc"]
-version = "0.4.0"
+version = "0.5.0"
```
Comment (ericphanson) [inline Project.toml:4]: (changing the defaults is breaking)

File: src/KeywordSearch.jl [hunk 2]
```diff
@@ -14,10 +14,13 @@ A list of replacements to automatically perform when preprocessing a [`Document`
 For example, if `KeywordSearch.AUTOMATIC_REPLACEMENTS == ["a" => "b"]`, then
 `Document("abc").text == "bbc"` instead of "abc".
 
-By default, `AUTOMATIC_REPLACEMENTS` contains only one replacement,
+By default, `AUTOMATIC_REPLACEMENTS` contains only one replacement:
+
+```jldoctest
+julia> KeywordSearch.AUTOMATIC_REPLACEMENTS
+1-element Vector{Pair{Union{Regex, String}, String}}:
+ r"[.!?><\-\n\r\v\t\f\s]+" => " "
```
Comment (ararslan) [inline src/KeywordSearch.jl]:
```suggestion
 r"[.!?><\-\v\f\s]+" => " "
```

Comment (ericphanson) [reply src/KeywordSearch.jl]: Reviewdog spotted this too, but you got there first! https://github.com/beacon-biosignals/KeywordSearch.jl/pull/23#discussion_r748609289

File: src/core.jl [hunk 3]
```diff
@@ -18,10 +18,17 @@ struct Document{T<:NamedTuple}
...
-        # as a word boundary.
-        new_text = " " * apply_replacements(text) * " "
+        # as a word boundary, if necessary.
+        if !startswith(new_text, " ") && !endswith(new_text, " ")
+            new_text = string(" ", new_text, " ")
+        elseif !startswith(new_text, " ")
+            new_text = string(" ", new_text)
+        elseif !endswith(new_text, " ")
+            new_text = string(new_text, " ")
+        end
```
Comment (ararslan) [inline src/core.jl:31]:
```suggestion
        if !startswith(new_text, " ")
            new_text = string(" ", new_text)
        end
        if !endswith(new_text, " ")
            new_text = string(new_text, " ")
        end
```

File: test/runtests.jl [hunk 4]
```diff
@@ -385,6 +385,25 @@ end
...
+
+        # Has tabs:
+        @test Document("   hello          goodbye   ??").text == " hello goodbye "
+        # Make sure we can match them too:
+        @test match(Query("hello goodbye"), Document("   hello     goodbye   ??")) !==
+              nothing
+        @test match(Query("hello ??? goodbye"), Document("   hello     goodbye   ??")) !==
+              nothing
+
+        # this has got tabs too
```
Comment (ararslan) [inline test/runtests.jl]: Does it? At least in the GitHub interface, it thinks all of the whitespace is spaces. You could use `\t` in the string to ensure there's a tab.

File: src/core.jl (hunk 3 above)
Comment (ericphanson) [reply src/core.jl:31]: I wanted to avoid allocating the string an extra time, in case it's a long one

Comment (ararslan) [reply src/core.jl:31]: Ah that's fair

File: test/runtests.jl (hunk 4 above)
Comment (ericphanson) [reply test/runtests.jl]: nope, my IDE normalized them out lol. Using `\t`!

**Review by ararslan — APPROVED** [review]

**Comment by ararslan** [comment]
(Well, approval pending CI passing :stuck_out_tongue:)

**Pull request merged**